pkg/dependency/dagmanager.go

```diff
@@ -0,0 +1,154 @@
+package dependency
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+// GlobalManager is the singleton instance of DAGManager.
+var GlobalManager DAGManager
+
+// DAGManager manages the dependency graphs (DAG) of all AppConfigs.
+// Each AppConfig has its own DAG where its components and traits depends on others unidirectionally.
+type DAGManager interface {
+	// Start is blocking call that bootstraps the manager code.
+	// It should be called in initialization stage.
+	Start(context.Context)
+
+	// AddDAG adds a dag of an AppConfig into DAGManager.
+	AddDAG(appKey string, dag *DAG)
+}
+
+// SetupGlobalDAGManager sets up the global dagManager.
+func SetupGlobalDAGManager(l logging.Logger, c client.Client) {
+	GlobalManager = &dagManagerImpl{
+		client:  c,
+		log:     l.WithValues("manager", "dag"),
+		app2dag: make(map[string]*DAG),
+	}
+}
+
+type dagManagerImpl struct {
+	mu     sync.Mutex
+	log    logging.Logger
+	client client.Client
+
+	app2dag map[string]*DAG
+}
+
+func (dm *dagManagerImpl) Start(ctx context.Context) {
+	for {
+		select {
+		case <-time.After(5 * time.Second):
+			dm.scan(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (dm *dagManagerImpl) scan(ctx context.Context) {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+
+	for app, dag := range dm.app2dag {
+		for sourceName, source := range dag.Sources {
+			// TODO: avoid repeating processing the same source by marking done in AppConfig status
+			val, err := dm.checkSourceReady(ctx, source)
+			if err != nil {
+				dm.log.Info("checkSourceReady failed", "errmsg", err)
+				continue
+			}
+			if val == "" { // not ready
+				continue
+			}
+
+			for sinkName, sink := range dag.Sinks[sourceName] {
+				dm.log.Debug("triggering sinks", "app", app, "source", sourceName, "sink", sinkName)
+				err := dm.trigger(ctx, sink, val)
+				if err != nil {
+					dm.log.Info("triggering sink failed", "errmsg", err)
+				}
+			}
+
+			delete(dag.Sources, sourceName)
```

Comment:
How can we delete source here when 
1. it's still in the loop
2. sinks are not guaranteed to be ready yet